### PR TITLE
feat: endpoint to logout app subdomain URLs

### LINKED ISSUE #4394: Logging out in the dashboard should also logout of subdomain app URLs
We could handle this by redirecting the user to a special `coder-app-logout` subdomain if `--wildcard-access-url` is set.

### PR DESCRIPTION
**Pull request by deansheather**
Closes #4394 

Adds a "special" app URL `coder-logout` (which is not a valid app URL so cannot be conflicted by real apps) which will delete the API key from the database and the cookie. Very lenient towards invalid tokens and will delete the cookie and redirect without failing even if it's an invalid token in most cases.

Accepts a `redirect_uri` query param which must have a hostname that matches the access URL or any app URL. Defaults to the access URL.

E.g. `https://coder-logout.dev.coder.com` would log you out of all app URLs and return you to the login page.

## TODO:
- [x] Change the frontend to redirect here (instead of login page) on logout

### CODE REVIEW

**Comment by deansheather** [comment]
Regarding the frontend, the implementation causes the login screen to flash before redirecting, but I'm unsure how to do it better in xstate.

**Review by f0ssel — APPROVED** [review]
good tests 👍

**Review by Kira-Pilot — APPROVED** [review]

**Comment by deansheather** [comment]
@Kira-Pilot is there a better way to redirect to the page without flashing the login screen for a split second?

**Review by kylecarbs — APPROVED** [review]

**Comment by deansheather** [comment]
This is blocked while we figure out a better way to avoid the double login page flash. Kira, Joe and I did some research into it in Discord on Friday and we found that the flash is caused by the `RequireAuth` component reacting to the state being "signedOut" and navigating to the login screen before we've finished navigating to the new logout endpoint.

The fix seems to be adding a new state to the auth state provider called "signedOutNavigatingAway" or something similar and having the `RequireAuth` component do nothing if it sees that state. Since Kira is out next week I might need some help from @BrunoQuaresma or @presleyp next week to get those changes finalized.

File: site/src/xServices/auth/authXService.ts [hunk 1]
```diff
@@ -469,7 +480,30 @@ export const authMachine =
...
+          // Get app hostname so we can see if we need to log out of app URLs.
+          // We need to load this before we log out of the API as this is an
+          // authenticated endpoint.
+          const appHost = await API.getApplicationsHost()
+          await API.logout()
+
+          if (appHost.host !== "") {
+            const { protocol, host } = window.location
+            const redirect_uri = encodeURIComponent(
+              `${protocol}://${host}/login`,
```
Comment (deansheather) [inline site/src/xServices/auth/authXService.ts:493]:
```suggestion
              `${protocol}//${host}/login`,
```

**Pull request merged**